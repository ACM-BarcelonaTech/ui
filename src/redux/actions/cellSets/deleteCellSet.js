import { CELL_SETS_DELETE } from '../../actionTypes/cellSets';
import saveCellSets from './saveCellSets';

const deleteCellSet = (experimentId, key) => async (dispatch, getState) => {
  const {
    loading, error,
  } = getState().cellSets;

  if (loading || error) {
    return null;
  }

  await dispatch({
    type: CELL_SETS_DELETE,
    payload: {
      experimentId,
      key,
    },
  });

  await dispatch(saveCellSets(experimentId));
};

export default deleteCellSet;
