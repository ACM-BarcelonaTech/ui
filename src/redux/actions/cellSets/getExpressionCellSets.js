import { fetchWork } from 'utils/work/fetchWork';
import pushNotificationMessage from 'utils/pushNotificationMessage';
import endUserMessages from 'utils/endUserMessages';

const getCellSetsByExpression = (experimentId, selectedGenes) => async (dispatch, getState) => {
  const body = {
    name: 'GetExpressionCellSets',
    genesConfig: selectedGenes,
  };

  try {
    const result = await fetchWork(experimentId, body, getState);
    return result;
  } catch (e) {
    console.error(e);
    pushNotificationMessage('error', endUserMessages.ERROR_FETCHING_CELL_SETS);
  }
};

export default getCellSetsByExpression;
