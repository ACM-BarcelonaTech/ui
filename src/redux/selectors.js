/* eslint-disable import/prefer-default-export */

// Accumulates all the different selectors

import * as backendSelectors from './selectors/backendStatus';
import * as cellSetsSelectors from './selectors/cellSets';

const getBackendStatus = (...params) => (state) => (
  backendSelectors.getBackendStatus(...params)(state.backendStatus));

const getCellSets = (...params) => (state) => {
  console.log('inside selector');
  return cellSetsSelectors.getCellSets(...params)(state.cellSets);
};

export {
  getBackendStatus,
  getCellSets,
};
