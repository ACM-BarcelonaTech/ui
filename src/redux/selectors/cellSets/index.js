import getCellSets from './getCellSets';
import getCellSetsHierarchy from './getCellSetsHierarchy';
import getCellSetsHierarchyByType from './getCellSetsHierarchyByType';
import getCellSetsHierarchyByKey from './getCellSetsHierarchyByKey';

// eslint-disable-next-line import/prefer-default-export
export {
  getCellSets,
  getCellSetsHierarchy,
  getCellSetsHierarchyByType,
  getCellSetsHierarchyByKey,
};
