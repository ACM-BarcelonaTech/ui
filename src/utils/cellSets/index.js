import composeTree from 'utils/cellSets/composeTree';
import getAllCells from 'utils/cellSets/getAllCells';
import getSampleCells from 'utils/cellSets/getSampleCells';
import { getCellSetKey, getRootKey } from 'utils/cellSets/getCellSetKeys';

export {
  composeTree,
  getAllCells,
  getSampleCells,
  getCellSetKey,
  getRootKey,
};
