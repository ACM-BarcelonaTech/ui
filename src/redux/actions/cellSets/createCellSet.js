import { v4 as uuidv4 } from 'uuid';

import fetchAPI from 'utils/fetchAPI';

import { CELL_SETS_CREATE } from 'redux/actionTypes/cellSets';

import endUserMessages from 'utils/endUserMessages';
import pushNotificationMessage from 'utils/pushNotificationMessage';
import { isServerError, throwIfRequestFailed } from 'utils/fetchErrors';

const createCellSetJsonMerger = (newCellSet, cellClassKey) => (
  [{
    $match: {
      query: `$[?(@.key == "${cellClassKey}")]`,
      value: {
        children: [
          {
            $insert: {
              index: '-',
              value: newCellSet,
            },
          },
        ],
      },
    },
  }]
);

const createCellSet = (experimentId, name, color, cellIds) => async (dispatch, getState) => {
  const {
    loading, error,
  } = getState().cellSets;

  if (loading || error) {
    return null;
  }

  const data = {
    key: uuidv4(),
    name,
    color,
    cellIds: new Set([...cellIds].map((id) => parseInt(id, 10))),
    type: 'cellSets',
  };

  if (data.cellIds.size === 0) {
    pushNotificationMessage('info', endUserMessages.EMPTY_CLUSTER_NOT_CREATED);
    return;
  }

  dispatch({
    type: CELL_SETS_CREATE,
    payload: { ...data },
  });

  const url = `/v1/experiments/${experimentId}/cellSets`;

  try {
    const dataForUpload = {
      ...data,
      cellIds,
    };

    const response = await fetchAPI(
      url,
      {
        method: 'PATCH',
        headers: {
          'Content-Type': 'application/boschni-json-merger+json',
        },
        body: JSON.stringify(
          createCellSetJsonMerger(dataForUpload, 'scratchpad'),
        ),
      },
    );

    const json = await response.json();
    throwIfRequestFailed(response, json, endUserMessages.ERROR_SAVING);

    pushNotificationMessage('info', endUserMessages.NEW_CLUSTER_CREATED);
  } catch (e) {
    if (!isServerError(e)) {
      console.error(`fetch ${url} error ${e.message}`);
    }

    pushNotificationMessage('error', endUserMessages.ERROR_SAVING);
  }
};

export default createCellSet;
