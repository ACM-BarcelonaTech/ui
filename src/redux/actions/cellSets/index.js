import createCellSet from './createCellSet';
import deleteCellSet from './deleteCellSet';
import resetCellSets from './resetCellSets';

import loadCellSets from './loadCellSets';
import saveCellSets from './saveCellSets';
import updateCellSetHierarchy from './updateCellSetHierarchy';
import updateCellSetProperty from './updateCellSetProperty';
import updateCellSetSelected from './updateCellSetSelected';

export {
  createCellSet,
  deleteCellSet,
  resetCellSets,
  loadCellSets,
  saveCellSets,
  updateCellSetHierarchy,
  updateCellSetSelected,
  updateCellSetProperty,
};
