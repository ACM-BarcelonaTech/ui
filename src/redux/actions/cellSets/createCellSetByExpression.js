import { fetchWork } from 'utils/work/fetchWork';
import pushNotificationMessage from 'utils/pushNotificationMessage';
import endUserMessages from 'utils/endUserMessages';
import WorkResponseError from 'utils/WorkResponseError';
import { createCellSet } from 'redux/actions/cellSets';

const createCellSetByExpression = (
  experimentId, selectedGenes,
) => async (dispatch, getState) => {
  const body = {
    name: 'GetExpressionCellSets',
    genesConfig: selectedGenes,
  };

  try {
    const result = await fetchWork(experimentId, body, getState);

    const { name, color, cellIds } = result;
    dispatch(createCellSet(
      experimentId, name, color, cellIds,
    ));

    return result;
  } catch (e) {
    let errorMessage = endUserMessages.ERROR_FETCHING_CELL_SETS;
    if (e instanceof WorkResponseError) errorMessage = e.userMessage;

    pushNotificationMessage('error', errorMessage);
    throw new Error(e);
  }
};

export default createCellSetByExpression;
