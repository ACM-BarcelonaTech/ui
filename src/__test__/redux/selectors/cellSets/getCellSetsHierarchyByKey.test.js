import { getCellSetsHierarchyByKeys } from 'redux/selectors';
import { mockCellSets1 } from '__test__/test-utils/cellSets.mock';

describe('Get cell sets hierarchy selector by key test', () => {
  it('returns correctly by one key', () => {
    const returnedHierarchy = getCellSetsHierarchyByKeys(['anotherTest'])(mockCellSets1);
    expect(returnedHierarchy).toMatchSnapshot();
  });
  it('returns correctly by more than one key', () => {
    const returnedHierarchy = getCellSetsHierarchyByKeys(['test', 'anotherTest'])(mockCellSets1);
    expect(returnedHierarchy).toMatchSnapshot();
  });
});
