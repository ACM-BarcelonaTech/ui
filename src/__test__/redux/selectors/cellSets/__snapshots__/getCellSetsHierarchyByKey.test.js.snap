// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Get cell sets hierarchy selector by key test returns correctly by more than one key 1`] = `Array []`;

exports[`Get cell sets hierarchy selector by key test returns correctly by one key 1`] = `Array []`;
